streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1516,6 +1472,31 @@ object ZStream extends ZStreamPlatformSpecific {
   final def bracket[R, E, A](acquire: ZIO[R, E, A])(release: A => ZIO[R, Nothing, _]): ZStream[R, E, A] =
     managed(ZManaged.make(acquire)(release))
 
+  final def combine[R, E, A, B, C](s1: ZStream[R, E, A], s2: ZStream[R, E, B], lc: Int = 2, rc: Int = 2)(
+    f0: (Queue[Take[E, A]], Queue[Take[E, B]], Queue[Take[E, C]]) => ZIO[R, E, Unit]
+  ): ZStream[R, E, C] =
+    new ZStream[R, E, C] {
+      def fold[R1 <: R, E1 >: E, C1 >: C, S]: Fold[R1, E1, C1, S] =
+        ZManaged.succeedLazy { (s, cont, f) =>
+          def loop(s: S, q: Queue[Take[E, C]]): ZIO[R1, E1, S] =
+            if (!cont(s)) ZIO.succeed(s)
+            else
+              q.take.flatMap {
+                case Take.End      => IO.succeed(s)
+                case Take.Fail(e)  => IO.halt(e)
+                case Take.Value(c) => f(s, c).flatMap(s => loop(s, q))
+              }
+
+          for {
+            left   <- s1.toQueue(lc)
+            right  <- s2.toQueue(rc)
+            dest   <- Queue.unbounded[Take[E, C]].toManaged(_.shutdown)
+            _      <- f0(left, right, dest).toManaged_
```

Comment:
Don't we need to fork here?

Reply:
Do we actually need an output queue? I haven't tested if all combinators work with this, but what if we do the following:
- `f0: (Queue[Take[E, A]], Queue[Take[E, B]]) => ZIO[R, E, Take[E, C]]`
- run `f0` with `ZStream.unfoldM` and `unTake`

Does that make sense?

Reply:
Let me give it a try.

Reply:
Yes we can.